File: src/ee/ads/GuardedInterstitialAd.cpp
#include "ee/ads/GuardedInterstitialAd.hpp"

#include <ee/ads/internal/AsyncHelper.hpp>
#include <ee/core/ObserverHandle.hpp>
#include <ee/coroutine/Task.hpp>

namespace ee {
namespace ads {
using Self = GuardedInterstitialAd;

Self::GuardedInterstitialAd(const std::shared_ptr<IInterstitialAd>& ad)
    : ad_(ad) {
    loading_ = false;
    loaded_ = false;
    displaying_ = false;

    handle_ = std::make_unique<ObserverHandle>();
    handle_->bind(*ad_).addObserver({
        .onLoaded =
            [this] {
                loaded_ = true;

                // Propagation.
                dispatchEvent([](auto&& observer) {
                    if (observer.onLoaded) {
                        observer.onLoaded();
                    }
                });
            },
        .onClicked =
            [this] {
                // Propagation.
                dispatchEvent([](auto&& observer) {
                    if (observer.onClicked) {
                        observer.onClicked();
                    }
                });
            },
    });
}

Self::~GuardedInterstitialAd() {}

void Self::destroy() {
    ad_->destroy();
    handle_->clear();
}

bool Self::isLoaded() const {
    return loaded_;
}

Task<bool> Self::load() {
    if (loaded_) {
        co_return true;
    }
    if (displaying_) {
        co_return false;
    }
    if (loading_) {
        // Waiting.
        co_return co_await ad_->load();
    }
    loading_ = true;
    auto result = co_await ad_->load();
    loading_ = false;
    co_return result;
}

Task<bool> Self::show() {
    if (not loaded_) {
        co_return false;
    }
    if (loading_) {
        co_return false;
    }
    if (displaying_) {
        // Waiting.
        co_return co_await ad_->show();
    }
    displaying_ = true;
    auto result = co_await ad_->show();
    displaying_ = false;
    if (not result) {
        // Failed to show, can use this ad again.
    } else {
        loaded_ = false;
    }
    co_return result;
}
} // namespace ads
} // namespace ee

File: unity/Runtime/Ads/Internal/GuardedInterstitialAd.cs
using System.Threading.Tasks;

namespace EE.Internal {
    public class GuardedInterstitialAd
        : ObserverManager<IInterstitialAdObserver>, IInterstitialAd {
        private readonly IInterstitialAd _ad;
        private bool _loading;
        private bool _displaying;
        private readonly ObserverHandle _handle;

        public GuardedInterstitialAd(IInterstitialAd ad) {
            _ad = ad;
            _handle = new ObserverHandle();
            _handle.Bind(ad).AddObserver(new IInterstitialAdObserver {
                OnLoaded = () => {
                    IsLoaded = true;
                    DispatchEvent(observer => observer.OnLoaded?.Invoke());
                },
                OnClicked = () => DispatchEvent(observer => observer.OnClicked?.Invoke())
            });
        }

        public void Destroy() {
            _ad.Destroy();
            _handle.Clear();
        }

        public bool IsLoaded { get; private set; }

        public async Task<bool> Load() {
            if (IsLoaded) {
                return true;
            }
            if (_displaying) {
                return false;
            }
            if (_loading) {
                // Waiting.
                return await _ad.Load();
            }
            _loading = true;
            var result = await _ad.Load();
            _loading = false;
            return result;
        }

        public async Task<bool> Show() {
            if (!IsLoaded) {
                return false;
            }
            if (_loading) {
                return false;
            }
            if (_displaying) {
                // Waiting.
                return await _ad.Show();
            }
            _displaying = true;
            var result = await _ad.Show();
            _displaying = false;
            if (!result) {
                // Failed to show, can use this ad again.
            } else {
                IsLoaded = false;
            }
            return result;
        }
    }
}